File: backend/app/services/finops_service.py
# backend/app/services/finops_service.py
from typing import Dict, Optional, Tuple, List, Any
from datetime import datetime
import logging

logger = logging.getLogger(__name__)

# Repository interface for multi-tenant token and compute unit tracking
class TenantUsageRepository:
    def __init__(self):
        # In-memory storage for tenant usage metrics and limits (backed by TimescaleDB/PostgreSQL in production)
        self._tenant_usage: Dict[str, Dict[str, float]] = {
            "tenant-sovereign-dgm": {"tokens_used": 1420500, "compute_units": 380.5, "burn_rate_aed_hr": 48.2, "requests_today": 1240},
            "tenant-dewa-grid": {"tokens_used": 890200, "compute_units": 210.0, "burn_rate_aed_hr": 29.5, "requests_today": 890},
            "tenant-rta-mobility": {"tokens_used": 540000, "compute_units": 140.2, "burn_rate_aed_hr": 18.0, "requests_today": 620},
            "tenant-enterprise-damac": {"tokens_used": 412000, "compute_units": 95.0, "burn_rate_aed_hr": 14.2, "requests_today": 430},
            "tenant-community-sandbox": {"tokens_used": 48000, "compute_units": 12.0, "burn_rate_aed_hr": 1.5, "requests_today": 85}
        }
        self._tenant_limits: Dict[str, Dict[str, Any]] = {
            "tenant-sovereign-dgm": {
                "name": "Dubai Government Media Office (DGM)",
                "tier": "sovereign",
                "max_tokens": 10000000,
                "max_compute_units": 2000.0,
                "budget_monthly_aed": 50000.0,
                "allowed_models": ["sovereign-pro", "gemini-2.5-pro", "gpt-4o", "dual-consensus"],
                "status": "ACTIVE",
                "authority_clearance": "DEFCON-1_SOVEREIGN"
            },
            "tenant-dewa-grid": {
                "name": "DEWA Smart Grid Intelligence",
                "tier": "sovereign",
                "max_tokens": 5000000,
                "max_compute_units": 1000.0,
                "budget_monthly_aed": 25000.0,
                "allowed_models": ["sovereign-pro", "gemini-2.5-pro", "gpt-4o"],
                "status": "ACTIVE",
                "authority_clearance": "CRITICAL_INFRASTRUCTURE"
            },
            "tenant-rta-mobility": {
                "name": "RTA Dubai Mobility Autonomous Fabric",
                "tier": "enterprise",
                "max_tokens": 3000000,
                "max_compute_units": 600.0,
                "budget_monthly_aed": 15000.0,
                "allowed_models": ["pro", "gemini-2.5-flash", "gpt-4o-mini"],
                "status": "ACTIVE",
                "authority_clearance": "CIVIC_ENTERPRISE"
            },
            "tenant-enterprise-damac": {
                "name": "DAMAC Strategic Twin Development",
                "tier": "enterprise",
                "max_tokens": 2000000,
                "max_compute_units": 400.0,
                "budget_monthly_aed": 10000.0,
                "allowed_models": ["pro", "flash", "gpt-4o-mini"],
                "status": "ACTIVE",
                "authority_clearance": "CIVIC_ENTERPRISE"
            },
            "tenant-community-sandbox": {
                "name": "Public Sandbox Developer Cohort",
                "tier": "free",
                "max_tokens": 100000,
                "max_compute_units": 25.0,
                "budget_monthly_aed": 250.0,
                "allowed_models": ["flash", "gemini-2.0-flash"],
                "status": "ACTIVE",
                "authority_clearance": "COMMUNITY_OPEN"
            }
        }

    async def get_current_usage(self, tenant_id: str) -> Dict[str, float]:
        """Query for current token/compute usage metrics."""
        return self._tenant_usage.get(tenant_id, {"tokens_used": 0.0, "compute_units": 0.0, "burn_rate_aed_hr": 0.0, "requests_today": 0})

    async def get_tenant_limits(self, tenant_id: str) -> Dict[str, Any]:
        """Query tenant subscription and quota limits."""
        return self._tenant_limits.get(tenant_id, {
            "name": f"Tenant {tenant_id}",
            "tier": "free",
            "max_tokens": 100000,
            "max_compute_units": 20.0,
            "budget_monthly_aed": 200.0,
            "allowed_models": ["flash"],
            "status": "ACTIVE",
            "authority_clearance": "COMMUNITY_OPEN"
        })

    async def record_usage(self, tenant_id: str, tokens: int, compute_units: float, cost_aed: float):
        """Atomically increment tenant usage counters."""
        if tenant_id not in self._tenant_usage:
            self._tenant_usage[tenant_id] = {"tokens_used": 0.0, "compute_units": 0.0, "burn_rate_aed_hr": 0.0, "requests_today": 0}
        self._tenant_usage[tenant_id]["tokens_used"] += tokens
        self._tenant_usage[tenant_id]["compute_units"] += compute_units
        self._tenant_usage[tenant_id]["requests_today"] += 1
        self._tenant_usage[tenant_id]["burn_rate_aed_hr"] = round(cost_aed * 60, 2)

    async def list_all_tenants(self) -> List[Dict[str, Any]]:
        """List all active tenant profiles with live usage vs quotas."""
        tenants = []
        for tid, limits in self._tenant_limits.items():
            usage = self._tenant_usage.get(tid, {"tokens_used": 0, "compute_units": 0, "burn_rate_aed_hr": 0, "requests_today": 0})
            token_burn_pct = round((usage.get("tokens_used", 0) / max(limits.get("max_tokens", 1), 1)) * 100, 1)
            spent_aed = round(usage.get("tokens_used", 0) * 0.000073, 2) # ~AED conversion
            tenants.append({
                "tenant_id": tid,
                "name": limits.get("name"),
                "tier": limits.get("tier"),
                "status": limits.get("status"),
                "authority_clearance": limits.get("authority_clearance"),
                "tokens_used": usage.get("tokens_used", 0),
                "max_tokens": limits.get("max_tokens", 0),
                "token_burn_pct": token_burn_pct,
                "compute_units": usage.get("compute_units", 0),
                "max_compute_units": limits.get("max_compute_units", 0),
                "budget_monthly_aed": limits.get("budget_monthly_aed", 0),
                "spent_aed": spent_aed,
                "burn_rate_aed_hr": usage.get("burn_rate_aed_hr", 0),
                "requests_today": usage.get("requests_today", 0),
                "allowed_models": limits.get("allowed_models", [])
            })
        return tenants


class FinOpsService:
    def __init__(self, usage_repo: TenantUsageRepository):
        self.usage_repo = usage_repo

    async def estimate_request_cost(self, prompt_length: int, endpoint: str) -> float:
        """
        Estimates the cost of an incoming request based on prompt length and endpoint type.
        1 token ≈ 4 characters. Blended baseline token cost = $0.00002.
        """
        estimated_tokens = max(1, prompt_length / 4)
        
        # Dual-Model Consensus endpoint performs parallel verification with 2.2x compute multiplier
        if "consensus" in endpoint or "reason" in endpoint:
            cost_per_token = 0.000044
        elif "simulation" in endpoint:
            cost_per_token = 0.000065
        else:
            cost_per_token = 0.00002

        return estimated_tokens * cost_per_token

    async def check_tenant_limits(self, tenant_id: str, estimated_cost: float) -> Tuple[bool, Optional[str]]:
        """
        Checks if the tenant has sufficient budget/limits for the estimated cost.
        Returns (is_allowed, error_message).
        """
        usage = await self.usage_repo.get_current_usage(tenant_id)
        limits = await self.usage_repo.get_tenant_limits(tenant_id)

        # Check token quota limits
        estimated_tokens = estimated_cost / 0.00002
        current_tokens = usage.get("tokens_used", 0)
        max_tokens = limits.get("max_tokens", float('inf'))

        if current_tokens + estimated_tokens > max_tokens:
            return False, f"Tenant token quota exceeded ({int(current_tokens):,} / {int(max_tokens):,} tokens used)."

        # Check compute quota limits
        if usage.get("compute_units", 0) >= limits.get("max_compute_units", float('inf')):
            return False, "Tenant compute capacity limit reached for current billing cycle."

        # Check authority status
        if limits.get("status") != "ACTIVE":
            return False, f"Tenant account is currently {limits.get('status')}. Access restricted."

        return True, None

    async def determine_model_route(self, tenant_id: str, prompt_length: int) -> str:
        """
        Determines which model tier to route the request to based on tenant tier, 
        prompt complexity, and authority separation policies.
        """
        limits = await self.usage_repo.get_tenant_limits(tenant_id)
        tenant_tier = limits.get("tier", "free")

        if tenant_tier == "sovereign":
            # Sovereign tier routes to dedicated high-capacity sovereign enclave
            return "sovereign-pro"
        elif tenant_tier == "enterprise":
            # Enterprise routes to Pro unless prompt length is massive (>12,000 chars)
            if prompt_length > 12000:
                return "flash"
            return "pro"
        else:
            # Free / Community always routes to high-speed Flash tier
            return "flash"

File: server/finopsService.ts
export interface TenantUsageMetrics {
  tokensUsed: number;
  computeUnits: number;
  burnRateAedHr: number;
  requestsToday: number;
}

export interface TenantLimitConfig {
  id: string;
  name: string;
  tier: 'free' | 'pro' | 'enterprise' | 'sovereign';
  maxTokens: number;
  maxComputeUnits: number;
  budgetMonthlyAed: number;
  allowedModels: string[];
  status: 'ACTIVE' | 'THROTTLED' | 'SUSPENDED';
  authorityClearance: 'COMMUNITY_OPEN' | 'CIVIC_ENTERPRISE' | 'CRITICAL_INFRASTRUCTURE' | 'DEFCON-1_SOVEREIGN';
}

export interface TenantFinOpsState extends TenantLimitConfig {
  tokensUsed: number;
  tokenBurnPct: number;
  computeUnits: number;
  spentAed: number;
  burnRateAedHr: number;
  requestsToday: number;
}

class TenantUsageRepository {
  private usageMap: Map<string, TenantUsageMetrics> = new Map([
    ["tenant-sovereign-dgm", { tokensUsed: 1420500, computeUnits: 380.5, burnRateAedHr: 48.2, requestsToday: 1240 }],
    ["tenant-dewa-grid", { tokensUsed: 890200, computeUnits: 210.0, burnRateAedHr: 29.5, requestsToday: 890 }],
    ["tenant-rta-mobility", { tokensUsed: 540000, computeUnits: 140.2, burnRateAedHr: 18.0, requestsToday: 620 }],
    ["tenant-enterprise-damac", { tokensUsed: 412000, computeUnits: 95.0, burnRateAedHr: 14.2, requestsToday: 430 }],
    ["tenant-community-sandbox", { tokensUsed: 48000, computeUnits: 12.0, burnRateAedHr: 1.5, requestsToday: 85 }]
  ]);

  private limitsMap: Map<string, TenantLimitConfig> = new Map([
    ["tenant-sovereign-dgm", {
      id: "tenant-sovereign-dgm",
      name: "Dubai Government Media Office (DGM)",
      tier: "sovereign",
      maxTokens: 10000000,
      maxComputeUnits: 2000.0,
      budgetMonthlyAed: 50000.0,
      allowedModels: ["sovereign-pro", "gemini-2.5-pro", "gpt-4o", "dual-consensus"],
      status: "ACTIVE",
      authorityClearance: "DEFCON-1_SOVEREIGN"
    }],
    ["tenant-dewa-grid", {
      id: "tenant-dewa-grid",
      name: "DEWA Smart Grid Intelligence",
      tier: "sovereign",
      maxTokens: 5000000,
      maxComputeUnits: 1000.0,
      budgetMonthlyAed: 25000.0,
      allowedModels: ["sovereign-pro", "gemini-2.5-pro", "gpt-4o"],
      status: "ACTIVE",
      authorityClearance: "CRITICAL_INFRASTRUCTURE"
    }],
    ["tenant-rta-mobility", {
      id: "tenant-rta-mobility",
      name: "RTA Dubai Mobility Autonomous Fabric",
      tier: "enterprise",
      maxTokens: 3000000,
      maxComputeUnits: 600.0,
      budgetMonthlyAed: 15000.0,
      allowedModels: ["pro", "gemini-2.5-flash", "gpt-4o-mini"],
      status: "ACTIVE",
      authorityClearance: "CIVIC_ENTERPRISE"
    }],
    ["tenant-enterprise-damac", {
      id: "tenant-enterprise-damac",
      name: "DAMAC Strategic Twin Development",
      tier: "enterprise",
      maxTokens: 2000000,
      maxComputeUnits: 400.0,
      budgetMonthlyAed: 10000.0,
      allowedModels: ["pro", "flash", "gpt-4o-mini"],
      status: "ACTIVE",
      authorityClearance: "CIVIC_ENTERPRISE"
    }],
    ["tenant-community-sandbox", {
      id: "tenant-community-sandbox",
      name: "Public Sandbox Developer Cohort",
      tier: "free",
      maxTokens: 100000,
      maxComputeUnits: 25.0,
      budgetMonthlyAed: 250.0,
      allowedModels: ["flash", "gemini-2.0-flash"],
      status: "ACTIVE",
      authorityClearance: "COMMUNITY_OPEN"
    }]
  ]);

  public getCurrentUsage(tenantId: string): TenantUsageMetrics {
    return this.usageMap.get(tenantId) || { tokensUsed: 0, computeUnits: 0, burnRateAedHr: 0, requestsToday: 0 };
  }

  public getTenantLimits(tenantId: string): TenantLimitConfig {
    return this.limitsMap.get(tenantId) || {
      id: tenantId,
      name: `Tenant ${tenantId}`,
      tier: 'free',
      maxTokens: 100000,
      maxComputeUnits: 20.0,
      budgetMonthlyAed: 200.0,
      allowedModels: ["flash"],
      status: "ACTIVE",
      authorityClearance: "COMMUNITY_OPEN"
    };
  }

  public recordUsage(tenantId: string, tokens: number, computeUnits: number, costAed: number): void {
    const current = this.getCurrentUsage(tenantId);
    this.usageMap.set(tenantId, {
      tokensUsed: current.tokensUsed + tokens,
      computeUnits: current.computeUnits + computeUnits,
      burnRateAedHr: Number((costAed * 60).toFixed(2)),
      requestsToday: current.requestsToday + 1
    });
  }

  public setTenantStatus(tenantId: string, status: 'ACTIVE' | 'THROTTLED' | 'SUSPENDED'): boolean {
    const limits = this.limitsMap.get(tenantId);
    if (!limits) return false;
    limits.status = status;
    this.limitsMap.set(tenantId, limits);
    return true;
  }

  public listAllTenants(): TenantFinOpsState[] {
    const results: TenantFinOpsState[] = [];
    for (const [id, limits] of this.limitsMap.entries()) {
      const usage = this.getCurrentUsage(id);
      const tokenBurnPct = Number(((usage.tokensUsed / Math.max(limits.maxTokens, 1)) * 100).toFixed(1));
      const spentAed = Number((usage.tokensUsed * 0.000073).toFixed(2));
      results.push({
        ...limits,
        tokensUsed: usage.tokensUsed,
        tokenBurnPct,
        computeUnits: usage.computeUnits,
        spentAed,
        burnRateAedHr: usage.burnRateAedHr,
        requestsToday: usage.requestsToday
      });
    }
    return results;
  }

  public getTimeseriesAnalytics() {
    const tenants = this.listAllTenants();
    const days = 30;
    const currentDay = 15; // Mid-month for 2026 August
    const history: Array<{
      day: number;
      date: string;
      isForecast: boolean;
      totalTokens: number;
      totalSpendAed: number;
      totalComputeUnits: number;
      tenantBreakdown: Record<string, { tokens: number; spendAed: number; compute: number }>;
    }> = [];

    let cumSpend = 0;
    let cumTokens = 0;
    let cumCompute = 0;

    for (let d = 1; d <= days; d++) {
      const isForecast = d > currentDay;
      const dateStr = `Aug ${d.toString().padStart(2, '0')}`;
      
      const dayFactor = 0.8 + (Math.sin(d / 3) * 0.25) + (d / 35);
      const dailyTokens = Math.round(115000 * dayFactor);
      const dailySpend = Number((dailyTokens * 0.000073).toFixed(2));
      const dailyCompute = Number((18.5 * dayFactor).toFixed(1));

      cumSpend = isForecast ? cumSpend + dailySpend : Number((dailySpend * d * 0.95).toFixed(2));
      cumTokens = isForecast ? cumTokens + dailyTokens : Math.round(dailyTokens * d * 0.95);
      cumCompute = isForecast ? cumCompute + dailyCompute : Number((dailyCompute * d * 0.9).toFixed(1));

      const breakdown: Record<string, { tokens: number; spendAed: number; compute: number }> = {};
      tenants.forEach(t => {
        const share = t.tier === 'sovereign' ? 0.45 : t.tier === 'enterprise' ? 0.22 : 0.05;
        breakdown[t.id] = {
          tokens: Math.round(dailyTokens * share),
          spendAed: Number((dailySpend * share).toFixed(2)),
          compute: Number((dailyCompute * share).toFixed(1))
        };
      });

      history.push({
        day: d,
        date: dateStr,
        isForecast,
        totalTokens: isForecast ? cumTokens : Math.min(cumTokens, 3300000),
        totalSpendAed: isForecast ? cumSpend : Math.min(cumSpend, 241.0),
        totalComputeUnits: isForecast ? cumCompute : Math.min(cumCompute, 837.7),
        tenantBreakdown: breakdown
      });
    }

    return history;
  }
}

export class FinOpsService {
  public usageRepo: TenantUsageRepository;

  constructor(usageRepo = new TenantUsageRepository()) {
    this.usageRepo = usageRepo;
  }

  public estimateRequestCost(promptLength: number, endpoint: string): number {
    const estimatedTokens = Math.max(1, promptLength / 4);
    let costPerToken = 0.00002;
    if (endpoint.includes("consensus") || endpoint.includes("reason")) {
      costPerToken = 0.000044;
    } else if (endpoint.includes("simulation")) {
      costPerToken = 0.000065;
    }
    return estimatedTokens * costPerToken;
  }

  public checkTenantLimits(tenantId: string, estimatedCost: number): { isAllowed: boolean; errorMessage: string | null } {
    const usage = this.usageRepo.getCurrentUsage(tenantId);
    const limits = this.usageRepo.getTenantLimits(tenantId);

    const estimatedTokens = estimatedCost / 0.00002;
    if (usage.tokensUsed + estimatedTokens > limits.maxTokens) {
      return {
        isAllowed: false,
        errorMessage: `Tenant token quota exceeded (${usage.tokensUsed.toLocaleString()} / ${limits.maxTokens.toLocaleString()} tokens used).`
      };
    }

    if (usage.computeUnits >= limits.maxComputeUnits) {
      return {
        isAllowed: false,
        errorMessage: `Tenant compute capacity limit reached for current billing cycle (${usage.computeUnits.toFixed(1)} / ${limits.maxComputeUnits.toFixed(1)} units).`
      };
    }

    if (limits.status !== "ACTIVE") {
      return {
        isAllowed: false,
        errorMessage: `Tenant status is '${limits.status}'. Access restricted.`
      };
    }

    return { isAllowed: true, errorMessage: null };
  }

  public determineModelRoute(tenantId: string, promptLength: number): string {
    const limits = this.usageRepo.getTenantLimits(tenantId);
    if (limits.tier === 'sovereign') {
      return 'sovereign-pro';
    } else if (limits.tier === 'enterprise') {
      if (promptLength > 12000) return 'flash';
      return 'pro';
    } else {
      return 'flash';
    }
  }
}

export const finopsService = new FinOpsService();
